src/speakers/builder.rs

```diff
@@ -0,0 +1,553 @@
+/// Some methods are only available when this types counterpart: `DeviceIsSet` is present.
+pub struct DeviceNotSet;
+
+/// Builder for configuring and opening speakers input streams.
+#[must_use]
+pub struct SpeakersBuilder<Device, Config, E = fn(cpal::StreamError)>
+where
+    E: FnMut(cpal::StreamError) + Send + Clone + 'static,
+{
+    device: Option<(cpal::Device, Vec<SupportedStreamConfigRange>)>,
+    config: Option<super::config::OutputConfig>,
+    error_callback: E,
+
+    device_set: PhantomData<Device>,
+    config_set: PhantomData<Config>,
+}
+
+impl<Device, Config, E> Debug for SpeakersBuilder<Device, Config, E>
+where
+    E: FnMut(cpal::StreamError) + Send + Clone + 'static,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SpeakersBuilder")
+            .field(
+                "device",
+                &self
+                    .device
+                    .as_ref()
+                    .map(|d| d.0.name().unwrap_or("unknown".to_string())),
+            )
+            .field("config", &self.config)
+            .finish()
+    }
+}
+
+impl Default for SpeakersBuilder<DeviceNotSet, ConfigNotSet> {
+    fn default() -> Self {
+        Self {
+            device: None,
+            config: None,
+            error_callback: default_error_callback,
+
+            device_set: PhantomData,
+            config_set: PhantomData,
+        }
+    }
+}
+
+fn default_error_callback(err: cpal::StreamError) {
+    #[cfg(feature = "tracing")]
+    tracing::error!("audio stream error: {err}");
+    #[cfg(not(feature = "tracing"))]
+    eprintln!("audio stream error: {err}");
+}
+
+impl SpeakersBuilder<DeviceNotSet, ConfigNotSet, fn(cpal::StreamError)> {
+    /// Creates a new speakers builder.
+    ///
+    /// # Example
+    /// ```no_run
+    /// let builder = rodio::speakers::SpeakersBuilder::new();
+    /// ```
+    pub fn new() -> SpeakersBuilder<DeviceNotSet, ConfigNotSet, fn(cpal::StreamError)> {
+        Self::default()
+    }
+}
+
+impl<Device, Config, E> SpeakersBuilder<Device, Config, E>
+where
+    E: FnMut(cpal::StreamError) + Send + Clone + 'static,
+{
+    /// Sets the input device to use.
+    ///
+    /// # Example
+    /// ```no_run
+    /// # use rodio::speakers::{SpeakersBuilder, available_outputs};
+    /// let input = available_outputs()?.remove(2);
+    /// let builder = SpeakersBuilder::new().device(input)?;
+    /// # Ok::<(), Box<dyn std::error::Error>>(())
+    /// ```
+    pub fn device(
+        &self,
+        device: impl Into<cpal::Device>,
+    ) -> Result<SpeakersBuilder<DeviceIsSet, ConfigNotSet, E>, Error> {
+        let device = device.into();
+        let supported_configs = device
+            .supported_output_configs()
+            .map_err(|source| Error::OutputConfigs {
+                source,
+                device_name: device.name().unwrap_or_else(|_| "unknown".to_string()),
```

Comment:
`Device::name` has been deprecated in favor of `Device::description().name()` in cpal v0.17. You could also choose to pretty-print `Device::description`.